panel/io/compile.py

```diff
@@ -0,0 +1,177 @@
+    the ReactiveESM components defined therein.
+
+    Arguments
+    ---------
+    path : str | os.PathLike
+        The path to the Python module.
+
+    Returns
+    -------
+    List of ReactiveESM components defined in the module.
+    """
+    path_obj = pathlib.Path(path)
+    source = path_obj.read_text(encoding='utf-8')
+    runner = CodeRunner(source, path, [])
+    module = runner.new_module()
+    runner.run(module)
+    components = []
+    for v in module.__dict__.values():
+        if isinstance(v, type) and issubclass(v, ReactiveESM) and not v.abstract:
+            v.__path__ = path_obj.parent.absolute()
+            components.append(v)
+    return components
+
+
+def extract_packages(esm_code: str) -> dict[str, str]:
+    """
+    Extracts package version definitions from ESM code.
+
+    Arguments
+    ---------
+    esm_code : str
+        The ESM code to search for package imports.
+
+    Returns
+    -------
+    Dictionary of packages and their versions.
+    """
+    # Regex pattern to match import statements with URLs starting with https
+    pattern = r"import\s+.*?\s+from\s+['\"](https:\/\/[^\/]+\/([^@\/]+)(?:@([\d\.\w-]+))?)['\"]"
+
+    # Find all matching import statements
+    matches = re.findall(pattern, esm_code)
+
+    packages = {}
+
+    for match in matches:
+        replace, package_name, version = match
+        # Add to dictionary
+        packages[package_name] = version
+        esm_code = esm_code.replace(replace, package_name)
+
+    return esm_code, packages
+
+
+def extract_dependencies(component: type[ReactiveESM]) -> tuple[str, dict[str, any]]:
+    """
+    Extracts dependencies from a ReactiveESM component by parsing its
+    importmap and the associated code and replaces URL import
+    specifiers with package imports.
+
+    Arguments
+    ---------
+    component: type[ReactiveESM]
+        The ReactiveESM component to extract a dependency definition from.
+
+    Returns
+    -------
+    code: str
+        Code where the URL imports have been replaced by package imports.
+    dependencies: dict[str, str]
+        A dictionary of package dependencies and their versions.
+    """
+    importmap = component._process_importmap()
+    if not importmap:
+        raise ValueError(
+            f'{component.__name__} did not declare an importmap '
+            'and cannot be compiled.'
+        )
+
+    # Extract dependencies and versions
+    dependencies = {}
+    for key, url in importmap["imports"].items():
+        match = re.search(r'@(\d+\.\d+\.\d+)', url)
+        if key.endswith('/'):
+            key = key[:-1]
+
+        if match:
+            dependencies[key] = f"^{match.group(1)}"
+        else:
+            dependencies[key] = "latest"
+
+    esm = component._render_esm(compiled=False)
+    if issubclass(component, ReactComponent):
+        esm += '\nimport * as React from "react"\nimport { createRoot } from "react-dom/client"\nexport default {render, React, createRoot}'
```

Comment:
```suggestion
        esm += '\nimport * as React from "react"\nimport { createRoot } from "react-dom/client"\nexport default {render, React, createRoot}'
```

Maybe use `os.linesep` here?